stored_file.php

```diff
@@ -0,0 +1,387 @@
+    const SIZE_KIB = 1024;
+
+    protected static $__supported_exts = array(
+
+        'gif',	// image
+        'png',	// image
+        'jpg',	// image
+        'jpeg',	// image
+        'doc',	// word document
+        'docx',	// word document
+        'xls',	// excel spreadsheet
+        'xlsx',	// excel spreadsheet
+        'ppt',	// powerpoint
+        'pptx',	// powerpoint
+        'rtf',	// rich text document
+        'pdf',	// portable document
+        'mp3',   // mp3 audio
+        'csv',   // csv data
+        'txt',   // text file
+        'zip',   // zip file
+
+    );
+
+    public $id;
+    public $extension;
+    public $mime;
+    public $filename;
+    public $source;
+
+    protected $__moved;
+
+    public function actual_filename()
+    {
+        if ($this->__moved)
+            return static::resolve_file($this->filename);
+        return $this->source;
+    }
+
+    public function has_supported_extension($supported_exts = null)
+    {
+        if (!$this->extension)
+            $this->extension = static::parse_extension($this->filename);
+
+        if (!$this->extension)
+            return false;
+
+        if ($supported_exts && is_array($supported_exts) && count($supported_exts))
+        {
+            return in_array($this->extension,
+                $supported_exts);
+        }
+
+        return in_array($this->extension,
+            static::$__supported_exts);
+    }
+
+    public function exists()
+    {
+        return is_file($this->actual_filename());
+    }
+
+    public function size()
+    {
+        return filesize($this->actual_filename());
+    }
+
+    public function human_size()
+    {
+        $size = $this->size();
+
+        if ($size > static::SIZE_GIB)
+        {
+            $size = ($size / static::SIZE_GIB);
+            return sprintf('%.2f GiB', $size);
+        }
+
+        if ($size > static::SIZE_MIB)
+        {
+            $size = ($size / static::SIZE_MIB);
+            return sprintf('%.2f MiB', $size);
+        }
+
+        $size = ($size / static::SIZE_KIB);
+        return sprintf('%.2f KiB', $size);
+    }
+
+    public function detect_mime()
+    {
+        if ($this->mime) return $this->mime;
+        $this->mime = File_Util::detect_mime($this->actual_filename());
+        return $this->mime;
+    }
+
+    public function generate_filename($ext = null)
+    {
+        if (!$ext) $ext = $this->extension;
+
+        $hash = md5_file($this->source);
+        $dir1 = substr($hash, 0, 2);
+        $dir2 = substr($hash, 2, 2);
+        $name = substr($hash, 4, 28);
+
+        $this->filename = build_path($dir1, $dir2,
+            sf('%s.%s', $name, $ext));
+        return $this->filename;
+    }
+
+    public function read()
+    {
+        return file_get_contents($this->actual_filename());
+    }
+
+    public function move($invoker_tag = null)
+    {
+        if ($this->__moved) return;
+
+        if (!$invoker_tag) $invoker_tag = md5(serialize(debug_backtrace(5)));
```

Comment:
Using `5` here is not clear. Why did you use 5 instead of the constants?

Reply:
It's just a representation of the idea, I know that there are better ways to prepare this code, in fact, I could use a private method just for this. The number 5 was chosen just to take into account 3 possible methods calls with arguments.